File: GithubCopilot_BP_PY/CWE-522/522_1b_BP.py
import hashlib

import os

import re

import logging

from typing import Tuple



# Configure logging

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')



def generate_salt() -> bytes:

    """Generate a random salt."""

    return os.urandom(16)



def hash_password(password: str, salt: bytes) -> bytes:

    """Hash a password with a given salt."""

    return hashlib.pbkdf2_hmac('sha256', password.encode(), salt, 100000)



def validate_password(password: str) -> bool:

    """Validate the password against the policy."""

    if len(password) < 8:

        logging.warning("Password must be at least 8 characters long.")

        return False

    if not re.search("[a-z]", password):

        logging.warning("Password must contain at least one lowercase letter.")

        return False

    if not re.search("[A-Z]", password):

        logging.warning("Password must contain at least one uppercase letter.")

        return False

    if not re.search("[0-9]", password):

        logging.warning("Password must contain at least one digit.")

        return False

    if not re.search(r"[!@#$%^&*(),.?\":{}|<>]", password):

        logging.warning("Password must contain at least one special character.")

        return False

    return True



def store_password(username: str, password_hash: bytes, salt: bytes):

    """Securely store the password hash and salt."""

    # This is a placeholder for storing the hashed password and salt.

    # In a real application, this should be stored in a secure database.

    logging.info(f"Storing password hash and salt for {username}")



def authenticate_user(username: str, password: str) -> bool:

    """Authenticate the user."""

    # This is a placeholder for retrieving the user's stored password hash and salt.

    # In a real application, this should involve querying the secure database.

    stored_password_hash, stored_salt = b'', b''  # Placeholder values

    password_hash = hash_password(password, stored_salt)

    return password_hash == stored_password_hash



def register_user(username: str, password: str) -> Tuple[bool, str]:

    """Register a new user with a username and password."""

    if not validate_password(password):

        return False, "Password does not meet the policy requirements."

    try:

        salt = generate_salt()

        password_hash = hash_password(password, salt)

        store_password(username, password_hash, salt)

        logging.info(f"User {username} registered successfully.")

        return True, "User registered successfully."

    except Exception as e:

        logging.error(f"Error registering user: {e}")

        return False, "Error registering user."



# Example usage

if __name__ == "__main__":

    username = input("Enter username: ")

    password = input("Enter password: ")

    success, message = register_user(username, password)

    print(message)

File: ChatGPT_BP_Java/CWE-522/522_1b_BP.java
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.Base64;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class RegistrationForm {

    private static final int SALT_LENGTH = 16; // Salt length in bytes
    private static final String PASSWORD_REGEX = "^(?=.*[0-9])(?=.*[a-zA-Z]).{8,}$"; // Password must be at least 8 characters with letters and numbers
    private static final String LOG_FILE = "registration_log.txt"; // Log file path

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter username: ");
        String username = scanner.nextLine();

        System.out.print("Enter password: ");
        String password = scanner.nextLine();

        if (!isValidPassword(password)) {
            System.out.println("Password does not meet policy requirements.");
            logError("Invalid password for user: " + username);
            return;
        }

        try {
            // Generate salt and hash the password
            byte[] salt = generateSalt();
            String hashedPassword = hashPassword(password, salt);

            // Securely store the username and hashed password
            storeUserData(username, hashedPassword, salt);
            System.out.println("Registration successful!");

        } catch (Exception e) {
            System.out.println("An error occurred during registration: " + e.getMessage());
            logError("Registration error for user: " + username + " - " + e.getMessage());
        }

        scanner.close();
    }

    private static boolean isValidPassword(String password) {
        Pattern pattern = Pattern.compile(PASSWORD_REGEX);
        Matcher matcher = pattern.matcher(password);
        return matcher.matches();
    }

    private static byte[] generateSalt() {
        SecureRandom sr = new SecureRandom();
        byte[] salt = new byte[SALT_LENGTH];
        sr.nextBytes(salt);
        return salt;
    }

    private static String hashPassword(String password, byte[] salt) throws NoSuchAlgorithmException {
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        md.update(salt);
        byte[] hashedPassword = md.digest(password.getBytes(StandardCharsets.UTF_8));
        return Base64.getEncoder().encodeToString(hashedPassword);
    }

    private static void storeUserData(String username, String hashedPassword, byte[] salt) {
        // In a real application, you'd store this data in a secure database
        System.out.println("Storing user data...");
        System.out.println("Username: " + username);
        System.out.println("Hashed Password: " + hashedPassword);
        System.out.println("Salt: " + Base64.getEncoder().encodeToString(salt));
        logSuccess("User registered successfully: " + username);
    }

    private static void logError(String message) {
        // Log errors (in a real application, write to a logging framework or file)
        System.err.println("ERROR: " + message);
    }

    private static void logSuccess(String message) {
        // Log success messages
        System.out.println("SUCCESS: " + message);
    }
}

